plotly/utils.py

```diff
@@ -119,8 +119,72 @@ class PlotlyJSONEncoder(json.JSONEncoder):
 
     See PlotlyJSONEncoder.default for more implementation information.
 
+    Additionally, this encoder overrides nan functionality so that 'Inf',
+    'NaN' and '-Inf' encode to 'null'. Which is stricter JSON than the Python
+    version.
+
     """
 
+    # we want stricter JSON, so convert NaN, Inf, -Inf --> 'null'
+    nan_str = inf_str = neg_inf_str = 'null'
+
+    # uses code from official python json.encoder module. Same licence applies.
```

Comment:
situation kinda stinks, I actually documented it because this same thing was confusing to me as well:

https://github.com/plotly/python-api/blob/master/plotly/utils.py#L135-150

We only get a crack at stuff if the JSONEncoder *can't* encode it. What's more, as is mentioned in the SO post, when `allow_nan=False`, we *also* don't get a crack at it in our encoder. bummer!

Here's an example:

```python
import json
import numpy

class Foo(object):
    pass

class NullEncoder(json.JSONEncoder):
    def encode_as_true(self, obj):
        return None
    
    def default(self, obj):
        return self.encode_as_true(obj)

json.dumps(Foo(), cls=NullEncoder)  # here, it works
# 'null'
json.dumps(numpy.ma.core.masked, cls=NullEncoder)  # and again here
# 'null'
json.dumps(float('NaN'), cls=NullEncoder)  # no shot! this doesn't work as expected!
# 'NaN'
```